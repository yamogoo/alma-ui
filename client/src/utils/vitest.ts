import type { VueWrapper } from "@vue/test-utils";

export const getTypedEmittedEvent = <V>(
  wrapper: VueWrapper,
  eventName: string,
) => {
  return wrapper.emitted(eventName) as V[][];
};
