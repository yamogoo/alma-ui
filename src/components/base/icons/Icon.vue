<script setup lang="ts">
import { defineAsyncComponent, computed } from "vue";

import type {
  IconName,
  IconStyle,
  IconWeight,
  IconColor,
  IconSize,
} from "./icons";
import Skeleton from "@/components/base/skeleton/Skeleton.vue";

const props = defineProps<Props>();

const symbol = computed(() => {
  const name = props.name;
  const style = props.style;
  const weight = props.weight;

  return defineAsyncComponent(async () => {
    return import(`../../../assets/icons/${name}_${style}_${weight}.svg`);
  });
});
</script>

<script lang="ts">
export interface Props {
  name: IconName;
  style: IconStyle;
  weight: IconWeight;
  color?: IconColor;
  size?: IconSize;
}
</script>

<template>
  <div
    class="icon"
    :class="[
      { [`icon_size-${size}`]: !!size, [`icon_color-${color}`]: !!color },
    ]"
    data-testid="icon"
  >
    <Suspense>
      <component v-if="symbol" :is="symbol" viewBox="0 0 24 24"></component>
      <template #fallback>
        <Skeleton></Skeleton>
      </template>
    </Suspense>
  </div>
</template>

<style lang="scss">
@use "sass:map";

@mixin defineSizes($map: $icon) {
  @each $size, $val in $map {
    &_size-#{$size} {
      @include box(px2rem(map.get($val, "size")));
    }
  }
}

@mixin defineThemes($names) {
  @each $name in $names {
    &_color-#{$name} {
      @include themify($themes) {
        fill: themed("label.#{$name}");
      }
    }
  }
}

.icon {
  @include box(auto, inherit);
  line-height: 0;
  fill: inherit;
  @extend %base-transition;

  @include defineSizes();
  @include defineThemes(primary primary-inversed secondary disabled accent);

  svg {
    @include box(auto, inherit);
    fill: inherit;
    @extend %base-transition;

    path {
      fill: inherit;
      @extend %base-transition;
    }
  }

  .skeleton {
    @include box(100%);
  }
}
</style>
