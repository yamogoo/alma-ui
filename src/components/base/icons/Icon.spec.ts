import { describe, expect, test, vi } from "vitest";
import { mount } from "@vue/test-utils";

import type { IconName, IconStyle, IconWeight } from "./icons";

import Icon from "./Icon.vue";

const icons: Array<IconName> = ["back", "check"];

const weightList: Array<IconWeight> = ["100", "200", "300", "400", "500"];

describe("Icon", () => {
  test.each(icons)("should render %s icon from the Icons Set", async (name) => {
    const _name = name as IconName;
    const _style: IconStyle = "outline";

    for (const weight of weightList) {
      const wrapper = mount(Icon, {
        props: {
          name: _name,
          weight: weight,
          style: _style,
        },
      });

      await vi.dynamicImportSettled();

      const res = wrapper.find("svg");

      expect(res.exists()).toBe(true);
      expect(res.exists()).toMatchSnapshot();
    }
  });
});
