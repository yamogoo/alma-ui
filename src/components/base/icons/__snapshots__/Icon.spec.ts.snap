// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Icon > should render back icon from the Icons Set 1`] = `true`;

exports[`Icon > should render back icon from the Icons Set 2`] = `true`;

exports[`Icon > should render back icon from the Icons Set 3`] = `true`;

exports[`Icon > should render back icon from the Icons Set 4`] = `true`;

exports[`Icon > should render back icon from the Icons Set 5`] = `true`;

exports[`Icon > should render check icon from the Icons Set 1`] = `true`;

exports[`Icon > should render check icon from the Icons Set 2`] = `true`;

exports[`Icon > should render check icon from the Icons Set 3`] = `true`;

exports[`Icon > should render check icon from the Icons Set 4`] = `true`;

exports[`Icon > should render check icon from the Icons Set 5`] = `true`;
